builder/test/simplify/simplify.cc

```diff
@@ -708,6 +708,95 @@ TEST(simplify, slice_of_crop) {
+    slinky::dim dims[2] = {{0, 20, 2}, {0, 0, 42}};
+    raw_buffer_ptr constant_buf = raw_buffer::make(2, 2, dims);
+    std::memset(constant_buf->base, constant_buf->size_bytes(), 0x17);
+
+    raw_buffer_ptr sliced_constant_buf = raw_buffer::make_copy(*constant_buf);
+    sliced_constant_buf->rank = 1;
+
+    ASSERT_THAT(simplify(constant_buffer::make(b1, constant_buf, slice_dim::make(b3, b1, 1, 0, body))),
+        matches(constant_buffer::make(b3, sliced_constant_buf, body)));
+  }
+
+  {
+    // Basic test. Slicing two dimensions.
+    slinky::dim dims[2] = {{0, 10, 1}, {0, 20, 10}};
+    raw_buffer_ptr constant_buf = raw_buffer::make(2, 1, dims);
+
+    raw_buffer_ptr sliced_buf = raw_buffer::make_copy(*constant_buf);
+    sliced_buf->slice(/*dim=*/1, /*at=*/10);
+    sliced_buf->slice(/*dim=*/0, /*at=*/5);
+
+    ASSERT_THAT(simplify(constant_buffer::make(b1, constant_buf, slice_buffer::make(b3, b1, {5, 10}, body))),
+        matches(constant_buffer::make(b3, sliced_buf, body)));
+  }
+
+  {
+    // Test out-of-bounds slicing.
+    slinky::dim dims[2] = {{0, 10, 1}, {0, 20, 1}};
+    raw_buffer_ptr constant_buf = raw_buffer::make(2, 1, dims);
+
+    raw_buffer_ptr sliced_buf = raw_buffer::make_copy(*constant_buf);
+    sliced_buf->slice(/*dim=*/1, /*at=*/21);
+    sliced_buf->slice(/*dim=*/0, /*at=*/5);
+    ASSERT_EQ(sliced_buf->base, nullptr);
+
+    ASSERT_THAT(simplify(constant_buffer::make(b1, constant_buf, slice_buffer::make(b3, b1, {5, 21}, body))),
+        matches(constant_buffer::make(b3, sliced_buf, body)));
+  }
+  {
+    // Slicing of a tensor of rank 3.
+    slinky::dim dims[3] = {{0, 1, 1}, {0, 1, 2}, {0, 1, 4}};
+    for (int d = 0; d < 3; ++d) {
+      {
+        // Slice a single dimension.
+        raw_buffer_ptr constant_buf = raw_buffer::make(3, 1, dims);
+        raw_buffer_ptr sliced_buf = raw_buffer::make_copy(*constant_buf);
+        sliced_buf->slice(d, 0);
+        ASSERT_THAT(simplify(constant_buffer::make(b1, constant_buf, slice_dim::make(b3, b1, d, 0, body))),
+            matches(constant_buffer::make(b3, sliced_buf, body)));
+      }
+      {
+        // Slice two dimensions.
+        std::vector<expr> at;
+        at.reserve(3);
+        for (int i = 0; i < 3; ++i) {
+          if (i == d) {
+            at.emplace_back();
+          } else {
+            at.emplace_back(0);
+          }
+        }
+        raw_buffer_ptr constant_buf = raw_buffer::make(3, 1, dims);
+        raw_buffer_ptr sliced_buf = raw_buffer::make_copy(*constant_buf);
+        for (int i = 2; i >= 0; --i) {
+          if (i != d) {
+            sliced_buf->slice(i, 0);
+          }
+        }
+        ASSERT_THAT(simplify(constant_buffer::make(b1, constant_buf, slice_buffer::make(b3, b1, at, body))),
+            matches(constant_buffer::make(b3, sliced_buf, body)));
+      }
+    }
+    {
+      // Slice all dimensions.
+      std::vector<expr> at(3, 0);
+      raw_buffer_ptr constant_buf = raw_buffer::make(3, 1, dims);
```

Comment:
Are these tests going to fail with msan enabled? I think they are using uninitialized memory and expecting comparisons to work (which they technically do, but I think msan will complain).

For these tests to be interesting, the memory should be filled with non-constant data.